### ISSUE TITLE
Runtime dependency on `@babel/runtime`? 

### ISSUE THREAD

**Issue by donmccurdy** [react-spring/cannon-es#23]
I'm looking into updating a couple packages' dependencies from a custom fork of `cannon.js` to `cannon-es`. When installing and running my tests, however, I'm seeing the error about `@babel/runtime` below. Looking at the [generated output on unpkg](https://unpkg.com/cannon-es@0.9.0/dist/index.cjs.js), I see the dependency on `@babel/runtime` as well. I'm not particularly familiar with using Babel directly, and it's not part of my package. Is that supposed to be a runtime dependency? I'm assuming not, but if so shouldn't it be declared in `dependencies` or `peerDependencies`?

The project I'm updating now is [three-to-cannon](https://github.com/donmccurdy/three-to-cannon), and the issue can be reproduced with `npm install && npm run dist && npm test` after cloning the repo. Note that it's a library, not an application: I declare `cannon-es` as a peer dependency rather than bundling it with the package, and I cannot transpile `cannon-es` on behalf of my users.

```js
> three-to-cannon@2.0.0 test /Users/donmccurdy/Documents/Projects/three-to-cannon
> node test/index.js

internal/modules/cjs/loader.js:775
    throw err;
    ^

Error: Cannot find module '@babel/runtime/helpers/possibleConstructorReturn'
Require stack:
- /Users/donmccurdy/Documents/Projects/three-to-cannon/node_modules/cannon-es/dist/index.cjs.js
- /Users/donmccurdy/Documents/Projects/three-to-cannon/dist/three-pathfinding.js
- /Users/donmccurdy/Documents/Projects/three-to-cannon/test/index.js
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:772:15)
    at Function.Module._load (internal/modules/cjs/loader.js:677:27)
    at Module.require (internal/modules/cjs/loader.js:830:19)
    at require (internal/modules/cjs/helpers.js:68:18)
    at Object.<anonymous> (/Users/donmccurdy/Documents/Projects/three-to-cannon/node_modules/cannon-es/dist/index.cjs.js:7:1)
    at Module._compile (internal/modules/cjs/loader.js:936:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:947:10)
    at Module.load (internal/modules/cjs/loader.js:790:32)
    at Function.Module._load (internal/modules/cjs/loader.js:703:12)
    at Module.require (internal/modules/cjs/loader.js:830:19) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/Users/donmccurdy/Documents/Projects/three-to-cannon/node_modules/cannon-es/dist/index.cjs.js',
    '/Users/donmccurdy/Documents/Projects/three-to-cannon/dist/three-pathfinding.js',
    '/Users/donmccurdy/Documents/Projects/three-to-cannon/test/index.js'
  ]
}
npm ERR! Test failed.  See above for more details.
```

**Comment by marcofugaro** [comment]
Yup, `@babel/runtime` wasn't in dependencies as [explained in the documentation](https://babeljs.io/docs/en/babel-plugin-transform-runtime).

However we don't need that plugin, since it only ends up saving ~100 bytes.

Should be fixed in #24, @donmccurdy you can check if it works for now using `react-spring/cannon-es#fix-babel` as a version in the package.json.

**Comment by codynova** [comment]
Released in `cannon-es@0.9.1`

**Comment by donmccurdy** [comment]
Thanks for the quick fix! :)

I did have one error after that change:

```
> three-to-cannon@2.0.0 test /Users/donmccurdy/Documents/Projects/three-to-cannon
> node test/index.js

/Users/donmccurdy/Documents/Projects/three-to-cannon/node_modules/cannon-es/dist/cannon-es.cjs.js:13030
  performance = {};
              ^

ReferenceError: performance is not defined
    at Object.<anonymous> (/Users/donmccurdy/Documents/Projects/three-to-cannon/node_modules/cannon-es/dist/cannon-es.cjs.js:13030:15)
    at Module._compile (internal/modules/cjs/loader.js:936:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:947:10)
    at Module.load (internal/modules/cjs/loader.js:790:32)
    at Function.Module._load (internal/modules/cjs/loader.js:703:12)
    at Module.require (internal/modules/cjs/loader.js:830:19)
    at require (internal/modules/cjs/helpers.js:68:18)
    at Object.<anonymous> (/Users/donmccurdy/Documents/Projects/three-to-cannon/dist/three-pathfinding.js:1:7)
    at Module._compile (internal/modules/cjs/loader.js:936:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:947:10)
npm ERR! Test failed.  See above for more details.
```

This is only in my test suite, which runs in node.js, and was easy enough to patch. If you have users who are actually using cannon-es serverside, they may need the same fix. I don't see the related code in this project, perhaps it's generated in the build step somehow.

```js
const {performance} = require('perf_hooks');
global.performance = performance;
```

**Comment by codynova** [comment]
I believe that's in the World class file, it's an older attempt to polyfill. I will take a look soon if Marco doesn't beat me to it.

**Comment by codynova** [comment]
Hey @donmccurdy, I'm finally getting around to this. What do you think is the best solution, considering you need to `require('perf_hooks')` in Node? I can see this as an option, but not sure if there is a better way:

```ts
// Check if global window object exists to understand runtime context (browser vs node)
const hasWindow = new Function('try {return this===window;}catch(e){return false;}')

// performance.now()
let performance = hasWindow() ? window.performance : global.performance || {}

if (!performance.now) {
  let nowOffset = Date.now()
  if (performance.timing && performance.timing.navigationStart) {
    nowOffset = performance.timing.navigationStart
  }
  performance.now = () => Date.now() - nowOffset
}
```

**Comment by donmccurdy** [comment]
I would try to avoid `eval` and `Function()`, which will break for users who need a Content Security Policy.

The answer may depend on your build setup. In some cases you can configure it to output both Node.js- and Browser-friendly builds, see [How to write a JavaScript package for both Node and the browser](https://nolanlawson.com/2017/01/09/how-to-write-a-javascript-package-for-both-node-and-the-browser/). This may be more trouble than you want for a profiling dependency.

You could also try just putting it in an `if()` block. Say:

```
let performance;
if (typeof window !== 'undefined') {
  performance = window.performance;
} else {
  performance = require('perf_hooks').performance;
}
```

As long as the user's build system doesn't bundle a `perf_hooks` implementation automatically (some bundlers do this for node.js dependencies, I think?), that shouldn't affect build size.

Finally, you could just require node.js users to provide a reference to the object when they enable profiling. Mildly inconvenient, but predictable, and unlikely to affect the majority of users.

**Comment by codynova** [comment]
> Finally, you could just require node.js users to provide a reference to the object when they enable profiling. Mildly inconvenient, but predictable, and unlikely to affect the majority of users.

Thank you for the suggestions. I'd honestly rather rely on the existing polyfill than breaking the isomorphic API parity. The conditional require should be fine, I think perf_hooks is pretty tiny.

Fixed in #35 and released in `cannon-es@0.11.0`. Please let me know if you see any other issues.

**Comment by codynova** [comment]
This solution (#35) causes an issue with webpack builds #37

I'm not sure of the most elegant isomorphic solution. As mentioned above, we can manually polyfill `performance` in Node environments, or change the `cannon-es` API for Node environments to require passing in the `perf_hooks` module.

@marcofugaro Do you have any ideas for this?

**Comment by marcofugaro** [comment]
Yes I've found a solution in #38.

It avoids us from using any polyfill.

**Comment by codynova** [comment]
Released fix in `cannon-es@0.13.0`
